postgresql: 16.4 -> 17.0

wolfgangwalther opened the pull request:
## Description of changes

[PostgreSQL 17 has been released](https://www.postgresql.org/docs/release/17.0/).

This PR adds the new major version and then updates the `postgresql` package to it. The module's default version is unchanged as discussed in https://github.com/NixOS/nixpkgs/pull/329611#discussion_r1697467876.

What I did so far:
- Built postgresql 17 with and without JIT on x86_64-linux, aarch64-darwin and with pkgsMusl.
- Built related nixos tests on linux.
- Updated some extensions breaking with the new version, where an update was available. There are more which currently still break, but no new version / fix is available, yet.
- Started a nixpkgs-review run and found one related build failure, so far: Kept `python3Packages.asyncpg` at v16. (still running)

I did all this by cherry-picking the commits here on master.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
postgresql: 16.4 -> 17.0

As discussed in [1] this change is decoupled from the major version
update for the NixOS postgresql module.

[1]: https://github.com/NixOS/nixpkgs/pull/329611/files#r1701393693

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pg_similarity.nix:
```diff
@@ -1,14 +1,14 @@
+{ stdenv, lib, fetchFromGitHub, gcc, postgresql, unstableGitUpdater }:
 
 stdenv.mkDerivation {
 
```
```suggestion
```

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/pg_similarity.nix:
```diff
@@ -1,14 +1,14 @@
 
   pname = "pg_similarity";
-  version = "1.0";
+  version = "1.0-unstable-2021-01-12";
   src = fetchFromGitHub {
```
```suggestion
  version = "1.0-unstable-2021-01-12";

  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -150,7 +152,8 @@ let
       # https://postgr.es/m/flat/427c7c25-e8e1-4fc5-a1fb-01ceff185e5b%40technowledgy.de
-      ++ lib.optionals (stdenv'.hostPlatform.isDarwin && atLeast "16") [ "LDFLAGS_EX_BE=-Wl,-export_dynamic" ];
+      ++ lib.optionals (stdenv'.hostPlatform.isDarwin && atLeast "16") [ "LDFLAGS_EX_BE=-Wl,-export_dynamic" ]
+      ++ lib.optionals (atLeast "17") [ "--without-perl" ];
```
Why do we have perl in nativeBuildInputs if we define here without perl?

SuperSandro2000 commented on pkgs/development/python-modules/asyncpg/default.nix:
Shouldn't this overwrite be in python-packages.nix?

Ma27 replied:
Hmm, is there some consensus on what's preferred?
I'm somewhat indifferent on that since I can see reasons for both styles (explicitness for what's required and less error-prone backports[1] vs more unneeded changes to a file and weird-looking code if you ever have a reason to use an older version (`postgresql_16 = postgresql_15;` or whatever)).

[1] If the default postgresql is older on stable, a naive backport will leave a broken package even though the attribute of the newer package is available just not the default. Running into this regularly with Go stuff.

SuperSandro2000 commented:
> I can't nearly complete a run of it, though, without crashing my machine.

That is not expected.

> I would love to have a way to only test direct reverse dependencies of `postgresql` - but I don't know how.

We could do a hydra jobset but how do you assess the situation outside of postgres extensions: Are there expected to be many breakages?

SuperSandro2000 commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -114,7 +115,8 @@ let
     ]
-      ++ lib.optionals jitSupport [ llvmPackages.llvm.dev nukeReferences ];
+      ++ lib.optionals jitSupport [ llvmPackages.llvm.dev nukeReferences ]
+      ++ lib.optionals (atLeast "17") [ bison flex perl docbook_xml_dtd_45 docbook-xsl-nons libxslt ];
```
See https://peter.eisentraut.org/blog/2024/08/13/the-new-postgresql-17-make-dist for some more details

Ma27 commented:
fwiw it's on my radar and I'll try to give a review soonish!

takeda commented:
Is this being held because some packages are being broken? Would it make sense to release 17.0 but still keep 16.x as the default, and resolving the issues when promoting the default from 16 to 17?

This would allow others to start using 17.0 and possibly also find any issues with it and other packages.

Ma27 reviewed (changes requested):
Currently fails for me with

```
postgresql> applying patch /nix/store/4gn6alxx0i0h61bdx1cfq217gqmc4q6x-b27622c90869aab63cfe22159a459c57768b0fa4.patch
postgresql> patching file src/test/regress/expected/date.out
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 3 out of 3 hunks ignored -- saving rejects to file src/test/regress/expected/date.out.rej
postgresql> patching file src/test/regress/expected/horology.out
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 9 out of 9 hunks ignored -- saving rejects to file src/test/regress/expected/horology.out.rej
postgresql> patching file src/test/regress/expected/timestamptz.out
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 13 out of 13 hunks ignored -- saving rejects to file src/test/regress/expected/timestamptz.out.rej
postgresql> patching file src/test/regress/pg_regress.c
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 1 out of 1 hunk ignored -- saving rejects to file src/test/regress/pg_regress.c.rej
postgresql> patching file src/test/regress/sql/horology.sql
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 1 out of 1 hunk ignored -- saving rejects to file src/test/regress/sql/horology.sql.rej
postgresql> patching file src/test/regress/sql/timestamptz.sql
postgresql> Reversed (or previously applied) patch detected!  Assume -R? [n] 
postgresql> Apply anyway? [n] 
postgresql> Skipping patch.
postgresql> 1 out of 1 hunk ignored -- saving rejects to file src/test/regress/sql/timestamptz.sql.rej
```

SuperSandro2000 commented:
I excluded the patch with the below diff but now disallowed reference complains about some -dev packages leaking into bin/postgres

```diff
diff --git a/pkgs/servers/sql/postgresql/generic.nix b/pkgs/servers/sql/postgresql/generic.nix
index 90306fa3f5e8..d709cab1243a 100644
--- a/pkgs/servers/sql/postgresql/generic.nix
+++ b/pkgs/servers/sql/postgresql/generic.nix
@@ -79,7 +79,8 @@
       disallowedRequisites = [
         stdenv'.cc
       ] ++ (
-        map lib.getDev (builtins.filter (drv: drv ? "dev") finalAttrs.buildInputs)
+        # map lib.getDev (builtins.filter (drv: drv ? "dev") finalAttrs.buildInputs)
+        []
       ) ++ lib.optionals jitSupport [
         llvmPackages.llvm.out
       ];
@@ -170,7 +171,7 @@
       })

       # TODO: Remove this with the next set of minor releases
-      (fetchpatch (
+      (if olderThan "16" then fetchpatch (
         if atLeast "14" then {
           url = "https://github.com/postgres/postgres/commit/b27622c90869aab63cfe22159a459c57768b0fa4.patch";
           hash = "sha256-7G+BkJULhyx6nlMEjClcr2PJg6awgymZHr2JgGhXanA=";
@@ -183,7 +184,7 @@
           url = "https://github.com/postgres/postgres/commit/205813da4c264d80db3c3215db199cc119e18369.patch";
           hash = "sha256-L8/ns/fxTh2ayfDQXtBIKaArFhMd+v86UxVFWQdmzUw=";
           excludes = [ "doc/*" ];
-        })
+        }) else {}
       )
     ] ++ lib.optionals stdenv'.hostPlatform.isMusl (
       # Using fetchurl instead of fetchpatch on purpose: https://github.com/NixOS/nixpkgs/issues/240141
```

Ma27 reviewed (commented):
Most of the stuff was reviewed already and the rest is also looking good.

wolfgangwalther committed:
python3Packages.asyncpg: 0.29.0 -> 0.30.0

This release supports PostgreSQL 17.

https://github.com/MagicStack/asyncpg/releases/tag/v0.30.0

kirillrdy commented:
sorry, i just realised that staging will not make it to 24.11 anyways, ignore my comment

Ma27 reviewed (changes requested):
According to #339153 there's a final staging-next cycle starting on Nov 6. This means that staging will be merged into staging-next then and we'd end up with the bump in 24.11. Hence blocking.

Code-wise it looks good and the touched packages build for me, so I'd merge right after that.
